blockscout-ens/bens-logic/src/subgraphs_reader/reader.rs

```diff
@@ -0,0 +1,233 @@
+use super::schema_selector::schema_names;
+use crate::{entity::subgraph::domain::Domain, hash_name::hash_ens_domain_name};
+use sqlx::postgres::PgPool;
+use std::{collections::HashMap, sync::Arc};
+use thiserror::Error;
+
+pub struct SubgraphReader {
+    pool: Arc<PgPool>,
+    schema_names: HashMap<i64, String>,
+}
+
+impl SubgraphReader {
+    pub async fn initialize(pool: Arc<PgPool>) -> Result<Self, anyhow::Error> {
+        let schema_names = schema_names(&pool).await?;
+        Ok(Self::new(pool, schema_names))
+    }
+
+    pub fn new(pool: Arc<PgPool>, schema_names: HashMap<i64, String>) -> Self {
+        Self { pool, schema_names }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum SubgraphReadError {
+    #[error("Network with id {0} not found")]
+    NetworkNotFound(i64),
+    #[error("Db err")]
+    DbErr(#[from] sqlx::Error),
+}
+
+impl SubgraphReader {
+    pub async fn get_domain(
+        &self,
+        network_id: i64,
+        name: &str,
+    ) -> Result<Option<Domain>, SubgraphReadError> {
+        let schema = self
+            .schema_names
+            .get(&network_id)
+            .ok_or_else(|| SubgraphReadError::NetworkNotFound(network_id))?;
+        let id = domain_id(name);
+        find_domain(self.pool.as_ref(), schema, &id).await
+    }
+
+    pub async fn get_domain_history(
+        &self,
+        _network_id: i64,
+        _name: &str,
+    ) -> Result<Vec<()>, SubgraphReadError> {
+        todo!()
+    }
+
+    pub async fn search_resolved_domain_reverse(
+        &self,
+        network_id: i64,
+        address: ethers::types::Address,
+    ) -> Result<Vec<Domain>, SubgraphReadError> {
+        let schema = self
+            .schema_names
+            .get(&network_id)
+            .ok_or_else(|| SubgraphReadError::NetworkNotFound(network_id))?;
+        let address = hex(address);
+        find_resolved_addresses(self.pool.as_ref(), schema, &address).await
+    }
+
+    pub async fn search_owned_domain_reverse(
+        &self,
+        network_id: i64,
+        address: ethers::types::Address,
+    ) -> Result<Vec<Domain>, SubgraphReadError> {
+        let schema = self
+            .schema_names
+            .get(&network_id)
+            .ok_or_else(|| SubgraphReadError::NetworkNotFound(network_id))?;
+        let address = hex(address);
+        find_owned_addresses(self.pool.as_ref(), schema, &address).await
+    }
+}
+
+async fn find_domain(
+    pool: &PgPool,
+    schema: &str,
+    id: &str,
+) -> Result<Option<Domain>, SubgraphReadError> {
+    let maybe_domain = sqlx::query_as(&format!(
+        r#"
+        SELECT *
+        FROM {schema}.domain
+        WHERE
+            id = $1 
+            AND name IS NOT NULL
+            AND (
+                expiry_date is null
+                OR to_timestamp(expiry_date) > now()
+            )
+            AND block_range @> 2147483647
```

Comment:
Nit: it make sense to add a comment here

Reply:
Moved common part to single variable and added comment for that var